### PR TITLE
Add support for snappy http content encoding

### PR DESCRIPTION
**Pull request by skyguard1**
Motivation:

Since netty already supports the decompression of snappy's http content encoding, it should also consider supporting snappy's http content compression

Modification:

Add support for snappy http content encoding

Result:

Netty supports snappy http content encoding

### CODE REVIEW

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java [hunk 1]
```diff
@@ -440,4 +460,16 @@ public MessageToByteEncoder<ByteBuf> createEncoder() {
                     zstdOptions.blockSize(), zstdOptions.maxEncodeSize());
         }
     }
+
+    /**
+     * Compression Encoder Factory for create {@link SnappyFrameEncoder}
+     * used to compress http content for snappy content encoding
+     */
+    private final class SnappyEncoderFactory implements CompressionEncoderFactory {
```
Comment (franz1981) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java:468]: static

Comment (skyguard1) [reply codec-http/src/main/java/io/netty/handler/codec/http/HttpContentCompressor.java]: Done

File: codec/src/main/java/io/netty/handler/codec/compression/SnappyOptions.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+/**
+ * {@link SnappyOptions} holds config for
+ * Snappy compression.
+ */
+public class SnappyOptions implements CompressionOptions {
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/SnappyOptions.java:22]:
```suggestion
public final class SnappyOptions implements CompressionOptions {
```

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DataCompressionHttp2Test.java [hunk 3]
```diff
@@ -338,6 +338,54 @@ public void run() throws Http2Exception {
         }
     }
 
+    @Test
+    public void snappyEncodingSingleEmptyMessage() throws Exception {
+        final String text = "";
+        final ByteBuf data = Unpooled.copiedBuffer(text.getBytes());
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DataCompressionHttp2Test.java:344]:
```suggestion
        final ByteBuf data = Unpooled.copiedBuffer(text.getBytes(CharsetUtil.US_ASCII));
```

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DataCompressionHttp2Test.java [hunk 4]
```diff
@@ -338,6 +338,54 @@ public void run() throws Http2Exception {
...
+            assertEquals(text, serverOut.toString(CharsetUtil.UTF_8.name()));
+        } finally {
+            data.release();
+        }
+    }
+
+    @Test
+    public void snappyEncodingSingleMessage() throws Exception {
+        final String text = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbbbbbbbbbbbbccccccccccccccccccccccc";
+        final ByteBuf data = Unpooled.copiedBuffer(text.getBytes(CharsetUtil.UTF_8.name()));
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DataCompressionHttp2Test.java:368]:
```suggestion
        final ByteBuf data = Unpooled.copiedBuffer(text.getBytes(CharsetUtil.US_ASCII)));
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec/src/main/java/io/netty/handler/codec/compression/SnappyOptions.java (hunk 2 above)
Comment (skyguard1) [reply codec/src/main/java/io/netty/handler/codec/compression/SnappyOptions.java]: Done

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DataCompressionHttp2Test.java (hunk 3 above)
Comment (skyguard1) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/DataCompressionHttp2Test.java]: Done

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DataCompressionHttp2Test.java (hunk 4 above)
Comment (skyguard1) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/DataCompressionHttp2Test.java]: Done

**Comment by normanmaurer** [comment]
@skyguard1 thanks a lot

**Pull request merged**